### PR TITLE
Support granular selection of tools to build

### PR DESCRIPTION
**Pull request by Qix-**
Ref #35.

This is one step closer to solving the issue; it allows you to disable/enable tools granularly.

By disabling `geometryc` and `texturev` and only building `shaderc` and `texturec`, I can now successfully pull in an external `meshoptimizer`, `stb`, `freetype2` and `harfbuzz` without any issues.

The reason for moving `cmake/tools.cmake`'s contents into the root was that if I didn't, I'd have to check _each_ of the tool's options in an `if()` statement, which would be prone to error if a tool was added later on (adding an option, putting it in `tools.cmake` but not including it in the if statement). It felt more correct.

The edit at the bottom of https://github.com/JoshuaBrookover/bgfx.cmake/issues/35#issuecomment-532594750 explains what needs to be done to have `geometryc` and `texturev` play nicely with parent projects, but those aren't covered by this PR.

### CODE REVIEW

**Comment by JonnyPtn** [comment]
Just for clarity this is almost completely unrelated to #35.

I don't see what this fixes or what the point of it is, but It also won't really break anything, so I don't feel strongly either way

**Comment by JoshuaBrookover** [comment]
I agree, it has nothing to do with the original issue in #35. It does relate to the discussion taking place in there. It's a good change, especially since texturev forces you to pull in example's dependencies.

**Comment by JoshuaBrookover** [comment]
Sorry, did not mean to close.

**Comment by JonnyPtn** [comment]
Probably worth adding a warning/error when trying to build texturev without the examples too so it’s clear?

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,16 +20,21 @@ if( APPLE AND NOT IOS )
...
-option( BGFX_AMALGAMATED      "Amalgamated bgfx build for faster compilation" OFF )
-option( BX_AMALGAMATED        "Amalgamated bx build for faster compilation"   OFF )
-option( BGFX_CONFIG_DEBUG     "Enables debug configuration on all builds"     OFF )
-option( BGFX_USE_DEBUG_SUFFIX "Add 'd' suffix to debug output targets"        ON  )
+# Disable BGFX_BUILD_TOOLS in order to specify specific tools to build.
+option( BGFX_BUILD_TOOLS          "Build all bgfx tools (overwrites all)."        ON  )
+option( BGFX_BUILD_TOOL_TEXTUREV  "Build bgfx tool: texturev"                     ${BGFX_BUILD_TOOLS} )
+option( BGFX_BUILD_TOOL_TEXTUREC  "Build bgfx tool: texturec"                     ${BGFX_BUILD_TOOLS} )
+option( BGFX_BUILD_TOOL_SHADERC   "Build bgfx tool: shaderc"                      ${BGFX_BUILD_TOOLS} )
+option( BGFX_BUILD_TOOL_GEOMETRYC "Build bgfx tool: geometryc"                    ${BGFX_BUILD_TOOLS} )
```
Comment (bwrsandman) [inline CMakeLists.txt:28]: I believe the idiomatic thing to do here is using `cmake_dependent_option`:
```cmake
cmake_dependent_option(BGFX_BUILD_TOOL_TEXTUREV "Build bgfx tool: texturev" ON
                       "BGFX_BUILD_TOOLS" OFF)
cmake_dependent_option(BGFX_BUILD_TOOL_TEXTUREC "Build bgfx tool: texturec" ON
                       "BGFX_BUILD_TOOLS" OFF)
cmake_dependent_option(BGFX_BUILD_TOOL_SHADERC "Build bgfx tool: shaderc" ON
                       "BGFX_BUILD_TOOLS" OFF)
cmake_dependent_option(BGFX_BUILD_TOOL_GEOMETRYC "Build bgfx tool: geometryc" ON
                       "BGFX_BUILD_TOOLS" OFF)
```
https://cmake.org/cmake/help/v3.15/module/CMakeDependentOption.html#module:CMakeDependentOption

Comment (Qix-) [reply CMakeLists.txt:28]: ~~We'd have to bump the minimum CMake version in order to use that. If @JoshuaBrookover is okay with that, then this is indeed the correct way to do this.~~

Disregard.

Comment (pezcode) [reply CMakeLists.txt:28]: It's supported by CMake 3.0 as well, no reason to bump version

**Comment by Qix-** [comment]
@pezcode good catch :)

File: CMakeLists.txt (hunk 1 above)
Comment (Qix-) [reply CMakeLists.txt:28]: Good catch @pezcode.

File: CMakeLists.txt [hunk 2]
```diff
@@ -57,11 +62,30 @@ include( cmake/bx.cmake )
 include( cmake/bimg.cmake )
 include( cmake/bgfx.cmake )
 
-if( BGFX_BUILD_TOOLS )
-	include( cmake/tools.cmake )
+if( BGFX_CUSTOM_TARGETS )
+	add_custom_target( tools )
```
Comment (JonnyPtn) [inline CMakeLists.txt:66]: What's the benefit of this custom target?

**Review by JonnyPtn — CHANGES_REQUESTED** [review]
You need to remove the submodule update from this PR

File: CMakeLists.txt (hunk 2 above)
Comment (JonnyPtn) [reply CMakeLists.txt:66]: Nevermind - was moved from deleted tools.cmake 👍

**Comment by Qix-** [comment]
@widberg why was this closed?

**Comment by widberg** [comment]
It is unrelated to the issue you referenced. You seem to be the only person that needs this. You changed the submodule remotes. If you want I’ll leave it open for others to comment, but the problem you are trying to solve does not appear to exist. I have no intention of merging this.

**Comment by Qix-** [comment]
> but the problem you are trying to solve does not appear to exist

Is nobody else using this in their project? Shaderc recompiles almost every time we do an incremental rebuild. Relinking it takes almost 20 seconds.

There's something very wrong with the cmake config as-is.

**Comment by bwrsandman** [comment]
It wasn't going to be merged with the changes to .gitmodules pointing to a different branch than upstream. Not to mention suggestions to fix the PR being ignored for months.
Your quick reply suggests that you are active on github.
I don't mind the issue you're trying to fix and if the PR is cleaned, then I'd very much like to see it merged. 

As for your rebuild issue, have you tried setting the target property EXCLUDE_FROM_ALL from your host projet? https://cmake.org/cmake/help/latest/prop_tgt/EXCLUDE_FROM_ALL.html#prop_tgt:EXCLUDE_FROM_ALL

**Comment by Qix-** [comment]
> It wasn't going to be merged with the changes to .gitmodules

Guys, I have already mentioned I understood this. This isn't what I'm arguing for.

> As for your rebuild issue, have you tried setting the target property EXCLUDE_FROM_ALL from your host projet?

No, `shaderc` is _not_ part of my own project. It's part of _this_ project.

I'm not going to respond to this, this is an innane discussion. I'll fork and maintain my own.

**Pull request closed without merging**